### PR TITLE
[Fix] Fix bugs and refactor codes in lora for better scalability.

### PR DESCRIPTION
**Pull request by aoshen524**
## Motivation

https://github.com/sgl-project/sglang/issues/3414 reports issues regarding limited model support compared to test_generation_models.py. This PR refines the LoRA dataclasses and tests several trending LoRA models on Hugging Face, uncovering some bugs and providing warnings for them to be addressed later.

## Modifications

- [x] Added an independent folder for lora-related tests.
- [x] Introduced LoRAAdaptor and LoRAModelCase dataclasses in utils.py for improved test case management in test_lora_backend.py.
- [x] Implemented dynamic tolerance settings based on the model and adaptor in tests in test_lora_backend.py.
- [x] Fixed an issue where some LoRA modules only had gate weights and no up weights by initializing the up weights to zero in lora.py.
- [x] Added a temporary restriction for not supporting embedding and LM head in lora_config.py.
- [x] Added skeleton for multi-LoRAbackend tests.
- [x] Fixed the handling of empty responses and ensured special tokens are skipped in runner.py.
- [x] Identified an accuracy problem when using flashinfer as the backend.

## Checklist

- [x] Add backend test support for single adaptor, single prompt inference.
- [ ] Add backend test support for single adaptor, batch prompts serving.
- [ ] Add backend test support for multi-adaptor, same rank.
- [ ] Add backend test support for multi-adaptor, different rank.
- [ ] Add backend test support for adaptor with Embedding and Lm_head layer weights.

### CODE REVIEW

File: test/srt/models/test_lora_backend.py [hunk 1]
```diff
@@ -43,141 +104,165 @@
...
-        for lora_set in LORA_SETS:
-            print(f"Testing lora set {lora_set}: ")
+
+        # Additional: compare prefill outputs between base model (no LoRA) and LoRA model for reference
+        hf_no_lora_prefill = torch.tensor(hf_no_lora_outputs.top_input_logprobs[0])
+        srt_no_lora_prefill = torch.tensor(srt_no_lora_outputs.top_input_logprobs[0])
+        print("Max diff (SRT base vs SRT LoRA prefill):", torch.max(torch.abs(srt_no_lora_prefill - srt_prefill)))
+        print("Max diff (HF base vs HF LoRA prefill):", torch.max(torch.abs(hf_no_lora_prefill - hf_prefill)))
+
+    def run_backend_batch(
```
Comment (Fridge003) [inline test/srt/models/test_lora_backend.py]: Maybe you can put the batch tests in another file? This file is mainly for testing the correctness of different backends.

File: test/srt/models/test_lora_backend.py [hunk 2]
```diff
@@ -12,22 +12,83 @@
...
-    {
-        "base": "meta-llama/Llama-3.1-8B-Instruct",
-        "loras": ["reissbaker/llama-3.1-8b-abliterated-lora"],
-        "decode_tolerance": 8e-2,
-    },
+from sglang.test.test_utils import calculate_rouge_l, is_in_ci
+
+
+@dataclasses.dataclass
+class LoRAAdaptor:
```
Comment (Fridge003) [inline test/srt/models/test_lora_backend.py]: Maybe the definition of LoRAAdapter and LoRAModelCase can be put in another file? So that other test files can also use them without repeated definition. Also a folder for lora tests can be created when necessary.

Comment (aoshen524) [reply test/srt/models/test_lora_backend.py:28]: sure

File: test/srt/models/test_lora_backend.py (hunk 1 above)
Comment (aoshen524) [reply test/srt/models/test_lora_backend.py:228]: I think batch tests like multi-lora support in checklist are also highly related to the correctness of the kernel in different backends. So I think yes. These should be in another file, but still need to be labeled as part of the lora backend tests.

File: test/srt/models/lora/utils.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import dataclasses
+from typing import List
+
+import torch
+
+
+@dataclasses.dataclass
+class LoRAAdaptor:
+    name: str
+    # 每个 adaptor 可单独配置容忍度（若未设置，则使用模型默认值）
```
Comment (Fridge003) [inline test/srt/models/lora/utils.py]: Please use English comments

**Review by Ying1123 — APPROVED** [review]
LGTM. Let's merge it after Baizhou's approval.

File: test/srt/models/lora/utils.py (hunk 3 above)
Comment (aoshen524) [reply test/srt/models/lora/utils.py]: ok

**Comment by Fridge003** [comment]
LGTM
